README.md
# Metrics Library for Go

[![Travis](https://img.shields.io/travis/istreamlabs/go-metrics.svg)](https://travis-ci.org/istreamlabs/go-metrics) [![Codecov](https://img.shields.io/codecov/c/github/istreamlabs/go-metrics.svg)](https://codecov.io/gh/istreamlabs/go-metrics) [![Documentation](https://img.shields.io/badge/godoc-reference-blue.svg)](https://godoc.org/github.com/istreamlabs/go-metrics/metrics) [![GitHub tag](https://img.shields.io/github/tag/istreamlabs/go-metrics.svg)](https://github.com/istreamlabs/go-metrics/releases) [![License](https://img.shields.io/github/license/istreamlabs/go-metrics.svg)](https://github.com/istreamlabs/go-metrics/blob/master/LICENSE)

A library that provides an interface for sending metrics from your script, application, or service. Metrics consist of a name, value, and optionally some tags that are made up of key/value pairs.

Multiple implementations are provided to allow for production, local development, and testing use cases. The following clients are available:

Client           | Description
---------------- | -----------
`LoggerClient`   | Writes metrics into a log stream. Useful when running locally.
`DataDogClient`  | Writes metrics into DataDog. Useful for production.
`NullClient`     | Acts like a mock that does nothing. Useful for testing.
`RecorderClient` | Writes metrics into memory and provides a query interface. Useful for testing.

## Example Usage

Generally you will instantiate one of the above clients and then write metrics to it. First, install the library:

```sh
# If using dep, glide, godep, etc then use that tool. For example:
dep ensure github.com/istreamlabs/go-metrics

# Otherwise, just go get it:
go get github.com/istreamlabs/go-metrics/metrics
```

Then you can use it:

```go
import "github.com/istreamlabs/go-metrics/metrics"

var client metrics.Client
if os.Getenv("env") == "prod" {
  client = metrics.NewDataDogClient("127.0.0.1:8125", "myprefix")
} else {
  // Log to standard out instead of sending production metrics.
  client = metrics.NewLoggerClient(nil)
}
defer client.Close()

// Simple incrementing counter
client.Incr("requests.count")

// Tagging with counters
client.WithTags(map[string]string{
  "tag": "value"
}).Incr("requests.count")
```

The above code would result in `myprefix.requests.count` with a value of `1` showing up in DataDog if you have [`dogstatsd`](https://docs.datadoghq.com/guides/dogstatsd/) running locally and an environment variable `env` set to `prod`, otherwise it will print metrics to standard out. See the [`Client`](https://godoc.org/github.com/istreamlabs/go-metrics/metrics/#Client) interface for a list of available metrics methods.

Sometimes you wouldn't want to send a metric every single time a piece of code is executed. This is supported by setting a sample rate:

```go
// Sample rate for high-throughput applications
client.WithRate(0.01).Incr("requests.count")
```

Sample rates apply to metrics but not events. Any count-type metric (`Incr`, `Decr`, `Count`, and timing/histogram counts) will get multiplied to the full value, while gauges are sent unmodified. For example, when emitting a 10% sampled timing metric that takes an average of `200ms` to DataDog, you would see `1 call * (1/0.1 sample rate) = 10 calls` added to the histogram count while the average value remains `200ms` in the DataDog UI.

Also provided are useful clients for testing. For example, the following asserts that a metric with the given name, value, and tag was emitted during a test:

```go
func TestFoo(t *testing.T)
  client := metrics.NewRecorderClient().WithTest(t)

  client.WithTags(map[string]string{
    "tag": "value",
  }).Count("requests.count", 1)

  // Now, assert that the metric was emitted.
  client.
    Expect("requests.count").
    Value(1).
    Tag("tag", "value")
}
```

For more information and examples, see the [godocs](https://godoc.org/github.com/istreamlabs/go-metrics/metrics).

## License

Copyright &copy; 2017 iStreamPlanet Co., LLC

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.8.0] - 2022-03-2

- Bump datadog-go dependency

## [1.7.0] - 2021-07-27

- Optimize tag serialization
- Bump datadog-go dependency versions

## [1.6.0] - 2021-06-10

- Add support for WithoutTelemetry option to configure DataDogClient to not send DD telemetry metrics.

## [1.5.0] - 2020-05-28

- Go module support
- Go 1.12-1.14 support
- Upgrade to datadog-go 3.x
- Adds support for DataDog Distributions: https://docs.datadoghq.com/metrics/distributions/
- Adds Close() to Client interface to support intentional flushes when applications are shutting down
- Optimize tag handling

## [1.4.0] - 2019-08-26

- Updated `datadog-go` to version `2.2.0`

## [1.3.1] - 2018-05-14

- Add more helpful error message when `WithTest(t)` is not called on the recorder metrics client during testing.

## [1.3.0] - 2018-03-19

- Add `WithRate(float64)` to the metrics interface and to all clients that implement
  the interface. All metrics calls support sample rates.
  - The `LoggerClient`:
    - Applies the sample rate when printing log messages. If the rate is `0.1` and you call `Incr()` ten times, expect about one message to have been printed out.
    - Displays the sample rate for counts if it is not `1.0`, e.g: `Count foo:0.2 (2 * 0.1) [tag1 tag2]`. This shows the sampled value, the passed value, and the sample rate.
    - Gauges, timings, and histograms will show the sample rate, but he value is left unmodified just like the DataDog implementation.
  - The `RecorderClient`:
    - Records all sample rates for metrics calls in `MetricCall.Rate`. No calls are excluded from the call list based on the sample rate, and the value recorded is the full value before multiplying by the sample rate.
    - Adds a `Rate(float64)` query method to filter by sampled metrics.
    - The following should work:

        ```go
        recorder := metrics.NewRecorderClient().WithTest(t)
        recorder.WithRate(0.1).Count("foo", 5)
        recorder.Expect("foo").Rate(0.1).Value(5)
        ```
- Add `Colorized()` method to `LoggerClient`, and automatically detect a TTY and enable color when `nil` is passed to the `NewLoggerClient` constructor.
- Test with Go 1.10.x.

## [1.2.0] - 2018-03-01

- Update build to test with Go 1.9, drop support for 1.7 since `dep` now
  requires Go 1.8+. Go 1.7 users can still use this library but must manage
  their own dependencies.
- Add `WithRate(rate float)` to the DataDog client to limit traffic sent to
  the `dogstatsd` daemon. The set rate will be applied to all calls made
  with the returned `Client`.
- Automatically assign tags to events.

## [1.1.0] - 2017-08-01

- Make `RecorderClient` goroutine-safe so that metrics can be written and
  checked concurrently. For example:

  ```go
  package main

  import (
    "sync"
    "github.com/istreamlabs/go-metrics/metrics"
  )

  func main() {
    client := metrics.NewRecorderClient()

    wg := sync.WaitGroup{}
    wg.Add(3)
    for i := 0; i < 3; i++ {
      go func() {
        client.Incr("concurrent.access")
        wg.Done()
      }()
    }
    wg.Wait()
  }
  ```

## [1.0.0] - 2017-07-26

- Make project public on GitHub.


metrics/client.go
// Package metrics provides a generic interface for sending metrics from your
// script, application, or service. Metrics consist of a name, a numeric value,
// and optionally a number of tags that are made up of key/value pairs.
//
// Multiple implementations are provided to allow for production, local
// development, and testing use cases. Generally you will instantiate one
// of these clients and then write metrics to it:
//
//   import "github.com/istreamlabs/go-metrics/metrics"
//
//   client := metrics.NewDataDogClient("127.0.0.1:8125", "myprefix")
//
//   // Simple incrementing counter
//   client.Incr("requests.count")
//
//   // Tagging with counters
//   client.WithTags(map[string]string{
//     "tag": "value"
//   }).Incr("requests.count")
//
//   // Sample rate for high-throughput applications
//   client.WithRate(0.01).Incr("requests.count")
//
// Also provided are useful clients for testing, both for when you want
// to assert that certain metrics are emitted and a `NullClient` for when
// you want to ignore them.
//
//   func TestFoo(t *testing.T) {
//     client := metrics.NewRecorderClient().WithTest(t)
//
//     client.WithTags(map[string]string{
//       "tag": "value",
//     }).Count("requests.count", 1)
//
//     // Now, assert that the metric was emitted.
//     client.
//       Expect("requests.count").
//       Value(1).
//       Tag("tag", "value")
//   }
//
package metrics

import (
	"time"

	"github.com/DataDog/datadog-go/v5/statsd"
)

// Client provides a generic interface to log metrics and events
type Client interface {
	// WithTags returns a new client with the given tags.
	WithTags(tags map[string]string) Client

	// WithRate returns a new client with the given sample rate.
	WithRate(rate float64) Client

	// Count/Incr/Decr set a numeric integer value.
	Count(name string, value int64)
	Incr(name string)
	Decr(name string)

	// Gauge sets a numeric floating point value.
	Gauge(name string, value float64)

	// Event creates a new event, which allows additional information to be
	// included when something worth calling out happens.
	Event(e *statsd.Event)

	// Timing creates a histogram of a duration.
	Timing(name string, value time.Duration)

	// Historgram creates a numeric floating point metric with min/max/avg/p95/etc.
	Histogram(name string, value float64)

	// Distribution tracks the statistical distribution of a set of values.
	Distribution(name string, value float64)

	// Close closes all client connections and flushes any buffered data.
	Close() error
}
